pysensors/optimizers/_gqr.py

```diff
@@ -0,0 +1,120 @@
+import numpy as np
+import pysensors
+
+from pysensors.optimizers._qr import QR
+
+import matplotlib.pyplot as plt
+from sklearn import datasets
+from sklearn import metrics
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+import pysensors as ps
+from matplotlib.patches import Circle
+from pysensors.utils._norm_calc import returnInstance as normCalcReturnInstance
+
+class GQR(QR):
+    """
+    General QR optimizer for sensor selection.
+    Ranks sensors in descending order of "importance" based on
+    reconstruction performance. This is an extension that requires a more intrusive
+    access to the QR optimizer to facilitate a more adaptive optimization. This is a generalized version of cost constraints
+    in the sense that users can allow n constrained sensors in the constrained area.
+    if n = 0 this converges to the CCQR results.
+
+    See the following reference for more information
+        Manohar, Krithika, et al.
+        "Data-driven sparse sensor placement for reconstruction:
+        Demonstrating the benefits of exploiting known patterns."
+        IEEE Control Systems Magazine 38.3 (2018): 63-86.
+
+    @ authors: Niharika Karnik (@nkarnik2999), Mohammad Abdo (@Jimmy-INL), and Krithika Manohar (@kmanohar)
+    """
+    def __init__(self):
+        """
+        Attributes
+        ----------
+        pivots_ : np.ndarray, shape [n_features]
+            Ranked list of sensor locations.
+        idx_constrained : np.ndarray, shape [No. of constrained locations]
+            Column Indices of the sensors in the constrained locations.
+        n_sensors : integer,
+            Total number of sensors
+        n_const_sensors : integer,
+            Total number of sensors required by the user in the constrained region.
+        all_sensors : np.ndarray, shape [n_features]
+            Optimall placed list of sensors obtained from QR pivoting algorithm.
+        constraint_option : string,
+            max_n_const_sensors : The number of sensors in the constrained region should be less than or equal to n_const_sensors.
+            exact_n_const_sensors : The number of sensors in the constrained region should be exactly equal to n_const_sensors.
+        """
+        self.pivots_ = None
+        self.idx_constrained = []
+        self.n_sensors = 10
+        self.n_const_sensors = 0
+        self.all_sensors = []
+        self.constraint_option = ''
+        self.nx = 64
```

Comment:
Is this a default option?

Reply:
default, feel free to suggest other defaults.

Reply:
We can remove the default for n_sensors as it is not there in QR or CCQR

Reply:
We can leave the nx and ny defaults and add a warning that it has defaulted to 64 pixels as we assume that they are using an image